packages/fleet/lib/supervisor.ts

```diff
@@ -0,0 +1,278 @@
+            states: ['PENDING', 'STARTING', 'RUNNING', 'OUTDATED', 'FINISHING', 'IDLE', 'TERMINATED', 'ERROR'],
+            ...(cursor ? { cursor } : {})
+        });
+        if (search.isErr()) {
+            return Err(search.error);
+        }
+        for (const [routingId, nodes] of search.value.nodes) {
+            // Start pending nodes
+            plan.push(...(nodes.PENDING || []).map((node) => ({ type: 'START' as const, node })));
+
+            // Timeout STARTING nodes if they are taking too long
+            plan.push(
+                ...(nodes.STARTING || []).flatMap((node) => {
+                    if (Date.now() - node.lastStateTransitionAt.getTime() > this.STATE_TIMEOUT_MS.STARTING) {
+                        return [{ type: 'FAIL' as const, node, reason: 'starting_timeout_reached' as const }];
+                    }
+                    return [];
+                })
+            );
+
+            // Mark OUTDATED nodes
+            plan.push(
+                ...(nodes.RUNNING || []).flatMap((node) => {
+                    if (node.deploymentId !== deployment.id) {
+                        return [{ type: 'OUTDATE' as const, node }];
+                    }
+                    return [];
+                })
+            );
+
+            // if OUTDATED node but no RUNNING or upcoming nodes then create a new one
+            if ((nodes.OUTDATED?.length || 0) > 0 && (nodes.RUNNING?.length || 0) + (nodes.STARTING?.length || 0) + (nodes.PENDING?.length || 0) === 0) {
+                plan.push({ type: 'CREATE' as const, routingId, deployment });
+            }
+
+            // Warn about old finishing nodes
+            plan.push(
+                ...(nodes.FINISHING || []).flatMap((node) => {
+                    if (Date.now() - node.lastStateTransitionAt.getTime() > this.STATE_TIMEOUT_MS.FINISHING) {
+                        return [{ type: 'WARN' as const, node }];
+                    }
+                    return [];
+                })
+            );
+
+            // Terminate IDLE nodes
+            plan.push(...(nodes.IDLE || []).map((node) => ({ type: 'TERMINATE' as const, node })));
+
+            // Remove old terminated nodes
+            plan.push(
+                ...(nodes.TERMINATED || []).flatMap((node) => {
+                    if (Date.now() - node.lastStateTransitionAt.getTime() > this.STATE_TIMEOUT_MS.TERMINATED) {
+                        return [{ type: 'REMOVE' as const, node }];
+                    }
+                    return [];
+                })
+            );
+
+            // Remove old error nodes
+            plan.push(
+                ...(nodes.ERROR || []).flatMap((node) => {
+                    if (Date.now() - node.lastStateTransitionAt.getTime() > this.STATE_TIMEOUT_MS.ERROR) {
+                        return [{ type: 'REMOVE' as const, node }];
+                    }
+                    return [];
+                })
+            );
+        }
```

Comment:
This plan strategy assume the supervisor is on top of everything, but machine could go down or up during this plan. Is there a plan for runners to reach out to the supervisor?

Reply:
runner would `register` (via jobs) when they start up to say they are online and ready.